connectivity/tests/encryption.go

```diff
@@ -54,53 +75,110 @@ func getInterNodeIface(ctx context.Context, t *check.Test, clientHost *check.Pod
 
 	device := strings.TrimRight(dev.String(), "\n\r")
 
-	if tunnelFeat, ok := t.Context().Feature(features.Tunnel); ok && tunnelFeat.Enabled {
+	if tunnelEnabled {
 		// When tunneling is enabled, and the traffic is routed to the cilium IP space
 		// we want to capture on the tunnel interface.
 		if device == defaults.HostDevice {
-			return "cilium_" + tunnelFeat.Mode // E.g. cilium_vxlan
+			return "cilium_" + tunnelMode // E.g. cilium_vxlan
 		}
 
 		// When both tunneling and host firewall is enabled, traffic from a pod
 		// to a remote nodes gets forwarded through the tunnel to preserve the
 		// source identity.
 		if hf, ok := t.Context().Feature(features.HostFirewall); ok && hf.Enabled &&
 			clientHost.Address(ipFam) != srcIP {
-			return "cilium_" + tunnelFeat.Mode // E.g. cilium_vxlan
+			return "cilium_" + tunnelMode // E.g. cilium_vxlan
 		}
 	}
 
 	return device
 }
 
-// getSourceAddressFilter constructs the source IP address filter we want to use for
-// capturing packet. If direct routing is used, the source IP is the client IP,
-// otherwise, either the client IP or the one associated with the egressing interface.
-func getSourceAddressFilter(ctx context.Context, t *check.Test, client,
-	clientHost *check.Pod, ipFam features.IPFamily, dstIP string,
-) string {
-	filter := fmt.Sprintf("src host %s", client.Address(ipFam))
-	if tunnelStatus, ok := t.Context().Feature(features.Tunnel); ok &&
-		!tunnelStatus.Enabled {
-		return filter
+// getFilter constructs a tcpdump filter to capture leakages of unencrypted pkts.
+//
+// The exact filter depends on the routing mode and some features. The common
+// structure is "src host $SRC_IP and dst host $DST_IP and $PROTO".
+//
+// WireGuard w/ tunnel: this is a special case. We are interested to see whether
+// any unencrypted VXLAN pkt is leaked on a native device.
+func getFilter(ctx context.Context, t *check.Test, client, clientHost *check.Pod,
+	server, serverHost *check.Pod,
+	ipFam features.IPFamily, reqType requestType) string {
+
+	tunnelEnabled := false
+	tunnelMode := ""
+	if tunnelStatus, ok := t.Context().Feature(features.Tunnel); ok && tunnelStatus.Enabled {
+		tunnelEnabled = true
+		tunnelMode = tunnelStatus.Mode
 	}
 
-	cmd := []string{
-		"/bin/sh", "-c",
-		fmt.Sprintf("ip -o route get %s | grep -oE 'src [^ ]*' | cut -d' ' -f2",
-			dstIP),
+	protoFilter := ""
+	switch reqType {
+	case requestHTTP:
+		protoFilter = "tcp"
+	case requestICMPEcho:
+		protoFilter = "icmp"
+		if ipFam == features.IPFamilyV6 {
+			protoFilter = "icmp6"
+		}
+	default:
+		t.Fatalf("Invalid request type: %d", reqType)
 	}
-	t.Debugf("Running %s", strings.Join(cmd, " "))
-	srcIP, err := clientHost.K8sClient.ExecInPod(ctx, clientHost.Pod.Namespace,
-		clientHost.Pod.Name, "", cmd)
-	if err != nil {
-		t.Fatalf("Failed to get IP route: %s", err)
+
+	if enc, ok := t.Context().Feature(features.EncryptionPod); tunnelEnabled && ok &&
```

Comment:
One drawback of using a different filter if wireguard is enabled is that we lose the sanity check, to prevent regressions in the test itself. I guess we cannot use this filter also for IPSec though, as in that case we encrypt then encapsulate, while with wireguard we encapsulate then encrypt IIUC.

Reply:
Yep. I listed this in a commit msg:

> Unfortunately, this makes the inverse tests (assertNoLeaks=false) noop
for this special case. But an additional check for leaks is going to be
added outside the CLI.

Reply:
Missed that one, sorry.